lib/granite/context_proxy.rb

```diff
@@ -0,0 +1,38 @@
+require 'granite/context_proxy/proxy'
+
+module Granite
+  # This concern contains class methods used for actions and projectors
+  #
+  module ContextProxy
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def using(context)
+        Proxy.new(self, context)
+      end
+
+      def as(performer)
+        using(performer: performer)
+      end
+
+      def with_context(context)
+        key = proxy_context_key
+        old_context = Thread.current[key]
+        Thread.current[key] = context
```

Comment:
Why do we need to store it in the fiber-local variable?

Reply:
Do you mean why fiber-local? To which I believe the answer is to support multithreaded servers (although I didn't write the code).
On the other hand if you mean why do we have to store it, honestly I think there might be a way to not store it, given that it's only used for the method chain. Not sure how yet, but perhaps you have some suggestions?

Reply:
Oh, now I see that it's from `ProxyPerformer` and I don't see an easy way to pass it without storing.
But do we need to store old context? Probably we can `Thread.current[key] = nil` in `ensure` and that's it, I don't see any way to use BA with `old_context`, since
```ruby
x = BA::Xxx.using(performer: a)
y = x.using(performer: b)
z = BA::Xxx.using(performer: c)
```
AFAIU will change context for all of them (which is quite confusing TBH).

Reply:
I think it's possible, with something like:

```ruby
class BA::Xxx
  def initialize
    @subaction = BA::Yyy.using(performer: random_person).new
  end
end

BA::Xxx.using(performer: different_person).new
```

In a very unlucky case where `BA::Xxx` and `BA::Yyy` instances have the same `hash` it would fail. Obviously very unlikely, but saving old context I think prevents this problem.

Reply:
Maybe `#object_id` then? `#hash` would mess up current context anyway in this scenario.

Reply:
Hm yeah `object_id` would actually work.

Reply:
```ruby
diff --git a/lib/granite/context_proxy.rb b/lib/granite/context_proxy.rb
index e5e72e7..4697f54 100644
--- a/lib/granite/context_proxy.rb
+++ b/lib/granite/context_proxy.rb
@@ -32,7 +32,7 @@ module Granite
       private
 
       def proxy_context_key
-        :"granite_proxy_performer_#{hash}"
+        :"granite_proxy_performer_#{object_id}"
       end
     end
   end
diff --git a/spec/lib/granite/context_proxy_spec.rb b/spec/lib/granite/context_proxy_spec.rb
index 838a384..1a723bc 100644
--- a/spec/lib/granite/context_proxy_spec.rb
+++ b/spec/lib/granite/context_proxy_spec.rb
@@ -76,11 +76,11 @@ RSpec.describe Granite::ContextProxy do
 
   describe '.proxy_context' do
     before do
-      Thread.current[:granite_proxy_performer_hash] = context
+      Thread.current[:"granite_proxy_performer_#{klass.object_id}"] = context
     end
 
     after do
-      Thread.current[:granite_proxy_performer_hash] = nil
+      Thread.current[:"granite_proxy_performer_#{klass.object_id}"] = nil
     end
 
     specify do
```
is green here /shrug

Reply:
I meant this as a spec to illustrate why `Thread.current[key] = nil` wouldn't work even if we used `object_id`.

Reply:
oh, ok. I didn't expected that `.with_context` is exposed as an public API.